From 0924ef6b28d46043984fe6a8bec2f6b9c5d472eb Mon Sep 17 00:00:00 2001
From: Jianzheng Zhou <jianzheng.zhou@freescale.com>
Date: Tue, 30 Jul 2013 13:27:29 +0800
Subject: [PATCH 1/2] ENGR00256988 wifi:load diff exec in start softap

For virtual device UNITE, it should start different hostapd daemon
according to differnt wifi vendor in starting softap.

Signed-off-by: Jianzheng Zhou <jianzheng.zhou@freescale.com>
---
 SoftapController.cpp | 17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/SoftapController.cpp b/SoftapController.cpp
index 825a376..a8aba3d 100644
--- a/SoftapController.cpp
+++ b/SoftapController.cpp
@@ -1,5 +1,6 @@
 /*
  * Copyright (C) 2008 The Android Open Source Project
+ * Copyright (C) 2013 Freescale Semiconductor, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -35,6 +36,7 @@
 
 #define LOG_TAG "SoftapController"
 #include <cutils/log.h>
+#include <cutils/properties.h>
 #include <netutils/ifc.h>
 #include <private/android_filesystem_config.h>
 #include "wifi.h"
@@ -44,6 +46,8 @@
 
 static const char HOSTAPD_CONF_FILE[]    = "/data/misc/wifi/hostapd.conf";
 static const char HOSTAPD_BIN_FILE[]    = "/system/bin/hostapd";
+static const char HOSTAPD_BIN_FILE_RTL[]    = "/system/bin/rtl_hostapd";
+static const char DRIVER_VENDOR_NAME[]  = "wlan.vendor";
 
 SoftapController::SoftapController()
     : mPid(0) {}
@@ -53,6 +57,7 @@ SoftapController::~SoftapController() {
 
 int SoftapController::startSoftap() {
     pid_t pid = 1;
+    char driver_vendor[PROPERTY_VALUE_MAX] = {'\0'};
 
     if (mPid) {
         ALOGE("SoftAP is already running");
@@ -66,9 +71,15 @@ int SoftapController::startSoftap() {
 
     if (!pid) {
         ensure_entropy_file_exists();
-        if (execl(HOSTAPD_BIN_FILE, HOSTAPD_BIN_FILE,
-                  "-e", WIFI_ENTROPY_FILE,
-                  HOSTAPD_CONF_FILE, (char *) NULL)) {
+        if ((property_get(DRIVER_VENDOR_NAME, driver_vendor, NULL)
+                    && (strcmp(driver_vendor, "realtek") == 0))?
+                (execl(HOSTAPD_BIN_FILE_RTL, HOSTAPD_BIN_FILE,
+                       "-e", WIFI_ENTROPY_FILE,
+                       HOSTAPD_CONF_FILE, (char *) NULL)):
+                (execl(HOSTAPD_BIN_FILE, HOSTAPD_BIN_FILE,
+                       "-e", WIFI_ENTROPY_FILE,
+                       HOSTAPD_CONF_FILE, (char *) NULL)
+                 )) {
             ALOGE("execl failed (%s)", strerror(errno));
         }
         ALOGE("SoftAP failed to start");
-- 
1.8.0

